File: apps/web/src/constants/globalBookingBrokerPromotionReasonCodes.ts
/**
 * Global booking broker promotion — aligned with ``shared/global_booking_broker_promotion_reason_codes.json``.
 * Write sets + operator_hint drive dropdowns; optional note is supplemental only.
 */
import spec from "@repo-shared/global_booking_broker_promotion_reason_codes.json";

type ReasonSpec = {
  schema_version: number;
  approve: { write: string[] };
  reject: { write: string[] };
  reopen: { write: string[] };
  operator_hint: Record<string, string>;
};

const data = spec as ReasonSpec;

export const GLOBAL_BROKER_PROMOTION_REASON_SCHEMA_VERSION = data.schema_version;

export type GlobalBrokerPromotionReasonOption = { code: string; label: string };

function optionsForCodes(codes: string[]): GlobalBrokerPromotionReasonOption[] {
  return codes.map((code) => ({
    code,
    label: data.operator_hint[code] ?? code,
  }));
}

/** Allowed reasons when moving to ``nextStatus`` from ``prevStatus`` (same rules as API). */
export function globalBrokerPromotionReasonOptions(prevStatus: string, nextStatus: string): GlobalBrokerPromotionReasonOption[] {
  const p = prevStatus.toLowerCase();
  const n = nextStatus.toLowerCase();
  if (n === "approved" && (p === "pending" || p === "rejected")) {
    return optionsForCodes(data.approve.write);
  }
  if (n === "rejected" && (p === "pending" || p === "approved")) {
    return optionsForCodes(data.reject.write);
  }
  if (n === "pending" && (p === "approved" || p === "rejected")) {
    return optionsForCodes(data.reopen.write);
  }
  return [];
}

export function globalBrokerPromotionTargets(prevStatus: string): { value: "approved" | "rejected" | "pending"; label: string }[] {
  const p = prevStatus.toLowerCase();
  if (p === "pending") {
    return [
      { value: "approved", label: "Approve" },
      { value: "rejected", label: "Reject" },
    ];
  }
  if (p === "approved") {
    return [
      { value: "rejected", label: "Reject" },
      { value: "pending", label: "Return to queue" },
    ];
  }
  if (p === "rejected") {
    return [
      { value: "approved", label: "Approve" },
      { value: "pending", label: "Return to queue" },
    ];
  }
  return [];
}

export function globalBrokerPromotionReasonHint(code: string | null | undefined): string | null {
  if (code == null || code === "") return null;
  const h = data.operator_hint[code];
  return h !== undefined ? h : null;
}

File: app/constants/global_booking_broker_promotion_reason_codes.py
"""Promotion ``reason_code`` values for global booking brokers — ``shared/global_booking_broker_promotion_reason_codes.json``.

* **approve / reject / reopen** ``write`` sets — allowed on ``PATCH`` when transitioning to that status.
* **operator_hint** — short UI labels (frontend imports the same JSON).

Legacy audit rows may still have older ``detail`` shapes (e.g. only ``note``); new changes always store
``promotion_reason_code`` plus ``promotion_reason_hint``.
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

_JSON_PATH = Path(__file__).resolve().parents[2] / "shared" / "global_booking_broker_promotion_reason_codes.json"


@lru_cache
def _data() -> dict[str, Any]:
    raw = _JSON_PATH.read_text(encoding="utf-8")
    return json.loads(raw)


def global_booking_broker_promotion_reason_schema_version() -> int:
    return int(_data()["schema_version"])


def _write_codes(group: str) -> frozenset[str]:
    return frozenset(str(x) for x in _data()[group]["write"])


APPROVE_PROMOTION_REASON_CODES_WRITE = _write_codes("approve")
REJECT_PROMOTION_REASON_CODES_WRITE = _write_codes("reject")
REOPEN_PROMOTION_REASON_CODES_WRITE = _write_codes("reopen")

ALL_PROMOTION_REASON_CODES: frozenset[str] = frozenset(
    APPROVE_PROMOTION_REASON_CODES_WRITE
    | REJECT_PROMOTION_REASON_CODES_WRITE
    | REOPEN_PROMOTION_REASON_CODES_WRITE
)


def promotion_operator_hint(code: str) -> str | None:
    hints = _data().get("operator_hint") or {}
    v = hints.get(code.strip())
    return str(v) if v is not None else None


def allowed_promotion_reasons_for_transition(*, prev_status: str, next_status: str) -> frozenset[str] | None:
    """Return allowed reason codes for ``prev_status`` → ``next_status``, or ``None`` if disallowed or no-op."""
    prev = (prev_status or "").strip().lower()
    nxt = (next_status or "").strip().lower()
    if prev == nxt:
        return None
    if nxt == "approved" and prev in ("pending", "rejected"):
        return APPROVE_PROMOTION_REASON_CODES_WRITE
    if nxt == "rejected" and prev in ("pending", "approved"):
        return REJECT_PROMOTION_REASON_CODES_WRITE
    if nxt == "pending" and prev in ("approved", "rejected"):
        return REOPEN_PROMOTION_REASON_CODES_WRITE
    return None


def normalize_and_validate_promotion_reason(
    *,
    prev_status: str,
    next_status: str,
    promotion_reason_code: str | None,
) -> tuple[str, frozenset[str]]:
    """Return ``(normalized_code, allowed_set)``. Raises ``ValueError`` with a stable message if invalid."""
    allowed = allowed_promotion_reasons_for_transition(prev_status=prev_status, next_status=next_status)
    if allowed is None:
        prev = (prev_status or "").strip().lower()
        nxt = (next_status or "").strip().lower()
        if prev == nxt:
            raise ValueError("global_broker_promotion_no_status_change")
        raise ValueError("global_broker_promotion_invalid_transition")

    raw = (promotion_reason_code or "").strip()
    if not raw:
        raise ValueError("global_broker_promotion_reason_required")
    if raw not in allowed:
        raise ValueError("global_broker_promotion_reason_not_allowed_for_transition")
    return raw, allowed
